File: packages/guards/src/response-quality.ts
import type { Guard, GuardContext, GuardResult } from 'open-guardrail-core';

interface ResponseQualityOptions {
  action: 'block' | 'warn';
  /** Minimum character count. Default: 10. */
  minLength?: number;
  /** Maximum ratio of repeated sentences. Default: 0.5. */
  maxRepetitionRatio?: number;
  /** Detect refusal/cop-out phrases. Default: true. */
  detectRefusal?: boolean;
}

const REFUSAL_PATTERNS = [
  /^I(?:'m| am) (?:sorry|afraid),? (?:but )?I (?:can'?t|cannot|am unable to)/i,
  /^As an AI(?: language model)?,? I (?:can'?t|cannot|don'?t)/i,
  /^I (?:do not|don'?t) have (?:the ability|access|enough)/i,
  /^Unfortunately,? I (?:can'?t|cannot|am unable)/i,
  /^I'?m not able to/i,
];

function computeRepetitionRatio(text: string): number {
  const sentences = text
    .split(/[.!?]+/)
    .map((s) => s.trim().toLowerCase())
    .filter((s) => s.length > 5);

  if (sentences.length < 2) return 0;

  const unique = new Set(sentences);
  return 1 - unique.size / sentences.length;
}

export function responseQuality(options: ResponseQualityOptions): Guard {
  const minLength = options.minLength ?? 10;
  const maxRepetitionRatio = options.maxRepetitionRatio ?? 0.5;
  const detectRefusal = options.detectRefusal ?? true;

  return {
    name: 'response-quality',
    version: '1.0.0',
    description: 'Check response quality: too short, repetitive, or refusal',
    category: 'content',
    supportedStages: ['output'],

    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      if (text.trim().length < minLength) {
        issues.push(`too-short (${text.trim().length} chars, min ${minLength})`);
      }

      const repRatio = computeRepetitionRatio(text);
      if (repRatio > maxRepetitionRatio) {
        issues.push(`repetitive (${(repRatio * 100).toFixed(0)}% duplicated sentences)`);
      }

      if (detectRefusal) {
        const firstLine = text.split('\n')[0];
        if (REFUSAL_PATTERNS.some((p) => p.test(firstLine))) {
          issues.push('refusal-detected');
        }
      }

      const detected = issues.length > 0;
      return {
        guardName: 'response-quality',
        passed: !detected,
        action: detected ? options.action : 'allow',
        score: detected ? Math.min(issues.length / 3, 1.0) : 0,
        message: detected ? `Quality issues: ${issues.join('; ')}` : undefined,
        details: detected ? { issues, repetitionRatio: repRatio } : undefined,
        latencyMs: Math.round(performance.now() - start),
      };
    },
  };
}

File: python/open_guardrail/guards/response_quality.py
"""Checks response quality heuristics."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult


class _ResponseQuality:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "response-quality"
        self.action = action

    def check(self, text: str, stage: str = "output") -> GuardResult:
        start = time.perf_counter()
        issues: List[str] = []
        if len(text.strip()) < 10:
            issues.append("Response too short (< 10 chars)")
        if text == text.upper() and len(text.strip()) > 0 and any(c.isalpha() for c in text):
            issues.append("Response is all caps")
        if re.search(r'[!?.]{11,}', text):
            issues.append("Excessive consecutive punctuation (> 10)")
        triggered = len(issues) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name=self.name,
            passed=not triggered,
            action=self.action if triggered else "allow",
            message="; ".join(issues) if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"issues": issues} if triggered else None,
        )


def response_quality(*, action: str = "warn") -> _ResponseQuality:
    return _ResponseQuality(action=action)
